### PR TITLE
Add support for providing CollationOptions

### PR DESCRIPTION
**Pull request by DarwinsBuddy**
Motivation:

Providing collation options for an index is available since mongo 3.4.
I thought vertx-mongoclient could use this feature as well, since the driver supports it.

### CODE REVIEW

**Comment by DarwinsBuddy** [comment]
@tsegismont It's also possible to provide collation for mongo's methods described in #264 according to the [documentation](https://docs.mongodb.com/manual/reference/collation/#operations-that-support-collation).
I stumbled upon this, when I tried to use collation with an index, so I implemented it for this only for now. If you think it's best to introduce the feature in all supporting methods at once, I'll have to take the effort, but this could take me a bit. :)

File: src/main/java/io/vertx/ext/mongo/CollationOptions.java [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+    }
+
+    @Override
+    public String toString() {
+      return this.value;
+    }
+  }
+
+  @SuppressWarnings("java:S115")
+  public enum MaxVariable {
```
Comment (tsegismont) [inline src/main/java/io/vertx/ext/mongo/CollationOptions.java:377]: These enums should be moved to top-level and be annotated with `@VertxGen`

File: src/main/java/io/vertx/ext/mongo/CollationOptions.java [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+      this.value = value;
+    }
+
+    @Override
+    public String toString() {
+      return this.value;
+    }
+  }
+
+  @SuppressWarnings("java:S115")
```
Comment (tsegismont) [inline src/main/java/io/vertx/ext/mongo/CollationOptions.java:376]: Please remove this

File: src/main/java/io/vertx/ext/mongo/CollationOptions.java [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+  boolean numericOrdering;
+  String alternate;
+  MaxVariable maxVariable;
+  boolean backwards;
+  boolean normalization;
+
+  /**
+   * Default constructor for setting
+   */
+  public CollationOptions(String locale) {
```
Comment (tsegismont) [inline src/main/java/io/vertx/ext/mongo/CollationOptions.java:38]: All Vert.x option classes must have at least 3 constructors: default (no param), copy and from json object.

File: src/main/java/io/vertx/ext/mongo/CollationOptions.java [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+  public static final MaxVariable DEFAULT_MAX_VARIABLE = MaxVariable.punct;
+  String locale;
+  boolean caseLevel;
+  CaseFirst caseFirst;
+  int strength;
+  boolean numericOrdering;
+  String alternate;
+  MaxVariable maxVariable;
+  boolean backwards;
+  boolean normalization;
```
Comment (tsegismont) [inline src/main/java/io/vertx/ext/mongo/CollationOptions.java:33]: Please make private

File: src/main/java/io/vertx/ext/mongo/CountOptions.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+  private String hintString;
+  private int limit;
+  private int skip;
+  private long maxTimeMS;
+  private CollationOptions collation;
+
+  public CountOptions() {
+  }
+
+  public CountOptions(JsonObject json) {
```
Comment (tsegismont) [inline src/main/java/io/vertx/ext/mongo/CountOptions.java:24]: Copy constructor is missing

File: src/main/java/io/vertx/ext/mongo/CountOptions.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+    return options
+      .hint(hint == null ? null : new JsonObjectBsonAdapter(hint))
+      .hintString(hintString)
+      .limit(limit)
+      .skip(skip)
+      .maxTime(maxTimeMS, TimeUnit.MILLISECONDS);
+  }
+
+  public JsonObject getHint() {
```
Comment (tsegismont) [inline src/main/java/io/vertx/ext/mongo/CountOptions.java:61]: Please add Javadoc for these properties

File: src/main/java/io/vertx/ext/mongo/CreateCollectionOptions.java [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+public class CreateCollectionOptions {
+  private long maxDocuments;
+  private boolean capped;
+  private long sizeInBytes;
+  private JsonObject storageEngineOptions = new JsonObject();
+  private IndexOptionDefaults indexOptionDefaults = new IndexOptionDefaults();
+  private ValidationOptions validationOptions = new ValidationOptions();
+  private CollationOptions collation;
+
+  public CreateCollectionOptions() {
```
Comment (tsegismont) [inline src/main/java/io/vertx/ext/mongo/CreateCollectionOptions.java:20]: Copy constructor is missing

File: src/main/java/io/vertx/ext/mongo/CreateCollectionOptions.java [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+  public boolean isCapped() {
+    return this.capped;
+  }
+
+  public CreateCollectionOptions setCapped(boolean capped) {
+    this.capped = capped;
+    return this;
+  }
+
+  public long getSizeInBytes() {
```
Comment (tsegismont) [inline src/main/java/io/vertx/ext/mongo/CreateCollectionOptions.java:62]: Please add Javadoc

File: src/main/java/io/vertx/ext/mongo/DistinctOptions.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
+package io.vertx.ext.mongo;
+
+import io.vertx.codegen.annotations.DataObject;
+import io.vertx.core.json.JsonObject;
+
+@DataObject(generateConverter = true)
+public class DistinctOptions {
```
Comment (tsegismont) [inline src/main/java/io/vertx/ext/mongo/DistinctOptions.java:7]: Same comments here: Javadoc + copy constructor

File: src/main/java/io/vertx/ext/mongo/IndexOptionDefaults.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
+package io.vertx.ext.mongo;
+
+import io.vertx.codegen.annotations.DataObject;
+import io.vertx.core.json.JsonObject;
+
+import java.util.Objects;
+
+@DataObject(generateConverter = true)
```
Comment (tsegismont) [inline src/main/java/io/vertx/ext/mongo/IndexOptionDefaults.java:8]: Same comments here: Javadoc + copy constructor

File: src/main/java/io/vertx/ext/mongo/IndexOptionDefaults.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+public class IndexOptionDefaults {
+  private JsonObject storageEngine = new JsonObject();
+
+  public IndexOptionDefaults() {
+    storageEngine = new JsonObject();
```
Comment (tsegismont) [inline src/main/java/io/vertx/ext/mongo/IndexOptionDefaults.java:13]: Is this the right name for the property? It looks like a copy/paste.

Also, why do we need a separate data object if the value is a single `JsonObject`?

File: src/main/java/io/vertx/ext/mongo/MongoClient.java [hunk 12]
```diff
@@ -75,25 +77,34 @@ static MongoClient createShared(Vertx vertx, JsonObject config) {
...
+   * @param settings       the native java mongo settings
    * @return the client
    */
   @GenIgnore
   static MongoClient createWithMongoSettings(Vertx vertx, JsonObject config, String dataSourceName, MongoClientSettings settings) {
     return new MongoClientImpl(vertx, config, dataSourceName, settings);
   }
 
+  @GenIgnore
+  static <T> DistinctPublisher<T> setDistinctOptions(DistinctPublisher<T> distinctPublisher, DistinctOptions distinctOptions) {
```
Comment (tsegismont) [inline src/main/java/io/vertx/ext/mongo/MongoClient.java:93]: This method should be moved to the impl class

File: src/main/java/io/vertx/ext/mongo/ValidationOptions.java [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+
+import com.mongodb.client.model.ValidationAction;
+import com.mongodb.client.model.ValidationLevel;
+import com.mongodb.lang.Nullable;
+import io.vertx.codegen.annotations.DataObject;
+import io.vertx.core.json.JsonObject;
+
+import java.util.Objects;
+
+@DataObject(generateConverter = true)
```
Comment (tsegismont) [inline src/main/java/io/vertx/ext/mongo/ValidationOptions.java:11]: Same comments: Javadoc + copy constructor

File: src/test/java/io/vertx/ext/mongo/DistinctTest.java [hunk 14]
```diff
@@ -306,4 +306,12 @@ public void testDistinctBatchWithQueryString() throws Exception {
     await();
     assertEquals(numDocs, results.size());
   }
+
+  @Test
+  public void testDistinctOptionsSetting() {
+    String collection = randomCollection();
+
+    mongoClient.distinctWithQuery(collection, "num", Integer.class.getName(), new JsonObject(), new DistinctOptions().setCollation(new CollationOptions("de_AT")));
```
Comment (tsegismont) [inline src/test/java/io/vertx/ext/mongo/DistinctTest.java:314]: It seems the expectations are not checked

**Review by tsegismont — CHANGES_REQUESTED** [review]
Thank you @DarwinsBuddy , another round of comments

**Comment by DarwinsBuddy** [comment]
Thanks for your feedback (it's my first vert.x contribution, so thanks for your patience)

I'll get back to you once I incorporated all of your comments.

**Comment by tsegismont** [comment]
No problem, thank *you* for your time and efforts

>

File: src/test/java/io/vertx/ext/mongo/MongoClientTest.java [hunk 15]
```diff
@@ -227,8 +255,10 @@ public void testAggregate() throws Exception {
 
   @Test
   public void testAggregateWithOptions() throws Exception {
+    CollationOptions collationOptions = new CollationOptions("de_AT");
```
[the remark this answers is not in the document]
Comment (DarwinsBuddy) [reply src/test/java/io/vertx/ext/mongo/MongoClientTest.java:258]: As `AggregateOptions` got enriched by optional `CoallationOptions` I thought setting it in the test case where the method for aggregating with Options is used, is straight forward of confirming that it's possible to add it. Though the testcase is not asserting the correct handling downwards to the driver, it's not needed or an additional test case for confirming the correct application of those options would be more beneficiary.

File: src/main/java/io/vertx/ext/mongo/IndexOptionDefaults.java (hunk 11 above)
Comment (DarwinsBuddy) [reply src/main/java/io/vertx/ext/mongo/IndexOptionDefaults.java:13]: As described here https://docs.mongodb.com/manual/reference/method/db.createCollection/

```
db.createCollection( <name>,
   {
     capped: <boolean>,
     timeseries: {                  // Added in MongoDB 5.0
        timeField: <string>,        // required for time series collections
        metaField: <string>,
        granularity: <string>
     },
     expireAfterSeconds: <number>,
     autoIndexId: <boolean>,
     size: <number>,
     max: <number>,
     storageEngine: <document>,
     validator: <document>,
     validationLevel: <string>,
     validationAction: <string>,
     indexOptionDefaults: <document>,
     viewOn: <string>,              // Added in MongoDB 3.4
     pipeline: <pipeline>,          // Added in MongoDB 3.4
     collation: <document>,         // Added in MongoDB 3.4
     writeConcern: <document>
   }
)
```
`indexOptionDefaults` is part of the options we can pass to a `createCollection` call. Adding now all of those fields and document their respective intentions and semantics are imho out of scope for the change, described in #256. Which doesn't mean, that we can not later on add these options to be passed to the mongo driver. Having this `DataObject` already in place eases the possibility to extend it later on, when needed.

File: src/main/java/io/vertx/ext/mongo/CollationOptions.java (hunk 3 above)
Comment (DarwinsBuddy) [reply src/main/java/io/vertx/ext/mongo/CollationOptions.java:38]: Unfortunately, I wasn't able to find the default locale set by mongo.
https://docs.mongodb.com/manual/reference/collation/
I fear that I'd change the intended behavior of setting a default locale without knowing that mongo also confirms to that.
Thus I decided to go for a mandatory locale, when using specific `CollationOptions`. @tsegismont do you know of a better solution or maybe which default locale I could set instead as a default one?

File: src/test/java/io/vertx/ext/mongo/MongoClientTest.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (tsegismont) [reply src/test/java/io/vertx/ext/mongo/MongoClientTest.java:258]: If the setting is not verified then it's better to remove it as we'll later wonder why it was added.

File: src/main/java/io/vertx/ext/mongo/CollationOptions.java (hunk 3 above)
Comment (tsegismont) [reply src/main/java/io/vertx/ext/mongo/CollationOptions.java:38]: If not locale is provided by the user, we should use the JVM's default locale, I think

File: src/main/java/io/vertx/ext/mongo/IndexOptionDefaults.java (hunk 11 above)
Comment (tsegismont) [reply src/main/java/io/vertx/ext/mongo/IndexOptionDefaults.java:13]: Looking at the Mongo documentation, it seems `storageEngine` should be part of `CreateCollectionOptions`, not nested in `IndexOptionsDefaults`.

**Comment by DarwinsBuddy** [comment]
@tsegismont I started incorporating your input, though I have some open questions regarding your feedback. Could you (whenever you have time) reply to the feedback I commented on?

The rest of the requested changes are reasonable and will be changed accordingly.

**Comment by tsegismont** [comment]
@DarwinsBuddy I thought I had replied this morning to every question you had. Which one did I miss?

File: src/main/java/io/vertx/ext/mongo/IndexOptionDefaults.java (hunk 11 above)
Comment (DarwinsBuddy) [reply src/main/java/io/vertx/ext/mongo/IndexOptionDefaults.java:13]: You're right. I am a sloppy reader. Thanks for clarifying

**Comment by DarwinsBuddy** [comment]
@tsegismont you did. I was awaiting a notification from GitHub, and was unsure if you'd get them. 🙈 Didn't realize that you already replied. Sorry for any inconveniences caused. I'll reply once I've finished up the PR

File: src/main/java/io/vertx/ext/mongo/IndexOptionDefaults.java (hunk 11 above)
Comment (DarwinsBuddy) [reply src/main/java/io/vertx/ext/mongo/IndexOptionDefaults.java:13]: After reconsidering the documentation it seems that `CreateCollectionOptions` takes `storageEngine` options at the root level of the options. But `indexOptionDefaults` takes another `storageEngine` object, which is used as default for created indexes on that collection. 
As the mongodb driver implements it like this

 https://github.com/mongodb/mongo-java-driver/blob/84910ac76988cebcd2730db77c167a4d685efceb/driver-core/src/main/com/mongodb/client/model/CreateCollectionOptions.java#L38

and

https://github.com/mongodb/mongo-java-driver/blob/84910ac76988cebcd2730db77c167a4d685efceb/driver-core/src/main/com/mongodb/client/model/IndexOptionDefaults.java#L30

I thought implementing a corresponding dataobject would be the straight forward way.

@tsegismont I could merely take a `JsonObject` as `IndexOptionDefaults` and simply look for `storageEngine` key, but on the other hand, having a corresponding dataObject class might be a cleaner solution. I'll change it as you wish. Just let me know how you decide :)

**Comment by DarwinsBuddy** [comment]
@tsegismont everything you mentioned should be fixed now with above commit, except https://github.com/vert-x3/vertx-mongo-client/pull/266#discussion_r760285425

I have not changed anything yet regarding this comment, but will wait for your reply and incorporate it accordingly.

File: src/main/java/io/vertx/ext/mongo/IndexOptionDefaults.java (hunk 11 above)
Comment (tsegismont) [reply src/main/java/io/vertx/ext/mongo/IndexOptionDefaults.java:16]: If storage engine is non structured data (it seems so reading the Mongo doc), then a mere `JsonObject` will do.

**Comment by DarwinsBuddy** [comment]
@tsegismont Thanks for your clarification. I replaced `IndexOptionDefaults` with a `JsonObject` and squashed the commits. I am sure you're busy releasing 4.2.2, though, if it's possible to already see this change in 4.2.2 I'd be very glad. But only if there is not too much effort involved. I can wait for 4.3.0 as well :) Have a nice day.

**Review by tsegismont — APPROVED** [review]
Thank you @DarwinsBuddy , this looks good to me

**Comment by vietj** [comment]
I did a couple of changes to this feature:

- the default collation locale is now "simple" because platform default locale is not suitable option and actually simple is the default locale when no collation is specified : https://github.com/vert-x3/vertx-mongo-client/commit/819a085a6c00698583ba49dcf3917212b1c2d57f
- a few enums are redundant with existing enum in mongo, so I removed them https://github.com/vert-x3/vertx-mongo-client/commit/79f8df70f3cec9aefd6a9ae15e5f93a44ecf4d72

**Pull request merged**